### PR TITLE
feat(state-keeper): implement asynchronous RocksDB cache

### PR DESCRIPTION
**Pull request by itegulov**
## What ❔

This PR adds a stateful storage abstraction `CachedStorage` that encapsulates both RocksDB and Postgres behind short-lived `Box<dyn StorageRead>` connections. Internally `CachedStorage` keeps track if RocksDB is behind Postgres and tries to catch it up in the background if so.

Apologies if the code is a little hairy, I tried to preserve the existing invariants as much as possible to limit the impact of this on other parts of the codebase.

Tested this locally by running load tests for 3 hours (~3k miniblocks), deleting RocksDB state and then restarting. Seems to be working fine, but it catches up really fast and I am not sure how to generate more data faster. So I am hoping to test this on testnet/staging with one of the external nodes.

Also trying to come up with a way to unit test this so that is WIP for now. In the meanwhile any feedback is welcome!

Note: metric for RocksDB-Postgres lag already [exists](https://github.com/matter-labs/zksync-era/blob/main/core/lib/state/src/rocksdb/metrics.rs#L13-L14)

## Why ❔

The main design goal here was to ensure liveliness of the system by providing a way to get a `ReadStorage` implementation while avoiding blocking operations as much as possible. If RocksDB ever gets wiped for whatever reason, State Keeper will keep working on Postgres-backed `StorageRead` connections for a few hours while `CachedStorage` is trying to catch RocksDB up.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `zk spellcheck`.
- [x] Linkcheck has been run via `zk linkcheck`.

### CODE REVIEW

File: core/lib/zksync_core/src/state_keeper/cached_storage.rs
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/state_keeper/cached_storage.rs:194]: why do we use postgres in this case? IIUC, we can use rocksDB

**Review by RomanBrodetski — COMMENTED** [review]
I need to look deeper, but I really like it from the first glance - it's very well abstracted from state keeper. 

I wanted to make a step back and think about `CachedStorageState` state transitions.
* Maybe naive, but do we really need "not initialized" status? can we start catching up and determine the correct state in `new` already? 
* Note that during state keeper operation, the cache **cannot fall behind by more than one block**. When state keeper is running, there is no other process that would be able to add blocks to L1 batches table. This is not the case on startup - on startup we might be way behind (reason (1) - SSD failed and we replaced it, reason (2) - we are recovering from snapshot). Thus, I think we don't need logic that would switch **from** rocksDB **to** Postgres - only the other way around. With this I think we can simplify the logic. 


I might definitely be off here - also curious what @slowli thinks

File: core/lib/zksync_core/src/state_keeper/cached_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/cached_storage.rs:18]: I certainly agree with @RomanBrodetski in that this looks overly complex. Conceptually, we want to represent two cases: either the storage has caught up (=> use RocksDB) or not (=> use Postgres). We can assume that the storage doesn't fall behind unless we let it to. So this should roughly correspond to

```rust
#[derive(Clone)]
enum CachedStorageFactory {
    Postgres(ConnectionPool),
    Rocksdb(RocksDB<StateKeeperColumnFamily>),
}

impl CachedStorageFactory {
    pub async fn access_storage(&self) -> Box<dyn ReadStorage + Send + '_> {
        // trivial. We may even not box the output type, but rather have an enum for it
        // (theoretically, can lead to better performance due to more
        // compiler optimizations being possible, so I'd go for it).
    }
}

pub struct CachedStorage {
    inner: Arc<Mutex<CachedStorageFactory>>,
    // ^ the background update task will replace `inner` once caught up.
}

impl CachedStorage {
    fn factory(&self) -> CachedStorageFactory {
         // it's important that we don't hold the lock for long;
         // that's why `CachedStorageFactory` implements `Clone`
         self.inner.lock().expect("poisoned").clone()
    }
}
```

- With this approach, an update task can be treated in the same way as other server / EN tasks; you wouldn't need to retain its handle locally.
- If you'd want to retain `ConnectionPool` in `CachedStorageFactory` even after it caught up (e.g., for it to be able to update to the latest L1 batch), this is possible as well by slightly changing its shape.

File: core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs:91]: - Nit: If you only use `RwLock::write()`, it makes sense to replace it with `Mutex`.
- In the general case, this would look like a *suspicious* use of `RwLock`: you lock it exclusively for an extended period of time. AFAIU, it works right now because L1 batches are processed strictly sequentially, but in general reducing the lock time to a minimum (like in the approach I've outlined in another comment) is always preferable.

File: core/lib/zksync_core/src/state_keeper/cached_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/cached_storage.rs:174]: Nit: You should add a lifetime here to make clear that `BoxReadStorage` borrows from `self`. This could make long-term `RwLock` locking more visible as well.

File: core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs
Comment (itegulov) [reply core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs]: Right, I think I was confused by lifetimes here and presumed that the guard gets dropped after `access_storage`. So your other comment about explicit lifetime makes a lot of sense!

File: core/lib/zksync_core/src/state_keeper/cached_storage.rs
Comment (itegulov) [reply core/lib/zksync_core/src/state_keeper/cached_storage.rs:21]: Indeed, I tried to account for the possibility that RocksDB somehow falls behind Postgres in the middle of the execution but if we don't care about that it simplifies a lot of things.

I have tried out your approach and it works perfectly. Just wondering what do you think I should do about errors? The existing codebase seems to be pretty split between:
a. Just unwrapping
b. Using `anyhow` for everything
c. Proper structured errors.

I feel like structured errors could make some sense here even if most underlying code is just `anyhow`. Also, representing interruption via an error would look cleaner than `None` as it is now (inherited approach from `RocksdbStorage`).

File: core/lib/zksync_core/src/state_keeper/cached_storage.rs
Comment (itegulov) [reply core/lib/zksync_core/src/state_keeper/cached_storage.rs]: I think I would rather leave this for a separate PR - this one is trying to do a lot of stuff as is

**Comment by itegulov** [comment]
@RomanBrodetski @slowli I have taken your comments into consideration and refactored the code according to them. Specifically, I am now operating under presumption that RocksDB can never fall more than one block behind after catching up which indeed simplified a lot of things.

I have also made the solution a bit more generic by introducing a trait `ReadStorageFactory` that regulates which `ReadStorage` implementations can be returned which is good for testing - I made so that we test some features specifically on Postgres-based storage or specifically on RocksDB-based storage. Also allowed me to write a test that can mimic RocksDB falling behind by executing a bunch of txs on Postgres first. Hopefully it did not make the PR too overengineered :)

PTAL when you can!

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:105]: Please use `<'_>` for `PgOrRocksdbStorage` to cleanly specify that it has a lifetime.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:157]: Nit: missing semicolon. Also, I wouldn't consider this log warranting WARN level; interruption is a part of the normal node function.

Also, completely subjectively, a let-else would read better here:

```rust
let Some(rocksdb) = rocksdb else {
    tracing::info!("Synchronizing RocksDB interrupted");
    return Ok(None);
};
// work with `rocksdb`
```

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:218]: Please define a type for the task so that it can be used in the [node framework](https://github.com/matter-labs/zksync-era/tree/main/core/node/node_framework) that @popzxc is developing. Also, in any case, you want to make this task cancelable (i.e., having it take `stop_receiver` rather than creating a surrogate one), and to propagate errors for it (so that if an errors occurs in it, it can be monitored and leads to the node termination). Maybe, we'd even want to have a [healthcheck](https://github.com/matter-labs/zksync-era/tree/main/core/lib/health_check) for this task to monitor syncing progress (naturally, this can be done in a follow-up PR). You can look at the API server as an example of the component with multiple tasks.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:20]: You could encapsulate `Arc<Mutex<_>>` wrapper inside `AsyncRocksdbCache` with something like

```rust
pub struct AsyncRocksdbCache(Arc<Mutex< AsyncRocksdbCacheInner>>);

enum AsyncRocksdbCacheInner {
    Postgres(ConnectionPool),
    Rocksdb(RocksDB<StateKeeperColumnFamily>, ConnectionPool),
}
```

One other thing you could try is to use [`OnceCell`](https://docs.rs/once_cell/latest/once_cell/sync/struct.OnceCell.html) to avoid a `Mutex` wrapper, so something like

```rust
struct AsyncRocksdbCache {
     pool: ConnectionPool,
     rocksdb: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
}
```

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:50]: Nit: You can use `Self::*` instead of `PgOrRocksdbStorage::*` in match clauses here and in other `ReadStorage` methods.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:47]: Nit: You can elide the lifetime here: `impl ReadStorage for PgOrRocksdbStorage<'_>`.

File: core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs
Comment (popzxc) [reply core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs:35]: `Arc<dyn Trait>` is our best friend now.

File: core/lib/state/src/utils.rs
Comment (slowli) [inline core/lib/state/src/utils.rs:13]: Bikeshedding: IMO, this change can be avoided; you can still retain `RocksdbStorage::builder()` method as a way to initialize the storage + convert the builder to `RocksDB<StateKeeperColumnFamily>` if required.

File: core/lib/zksync_core/src/state_keeper/batch_executor/tests/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/batch_executor/tests/mod.rs:54]: Bikeshedding: the methods in `StorageType` look like they belong to `Tester`, since they mostly operate on it, if this makes sense.

File: core/lib/zksync_core/src/state_keeper/batch_executor/tests/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/batch_executor/tests/mod.rs:158]: Nit: There's [`Product`](https://docs.rs/test-casing/latest/test_casing/struct.Product.html) in `test-casing` that provides the same functionality without a need to use another crate.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:133]: Nit: IMO, it wouldn't be controversial to retrieve a `Handle` here via [`Handle::current()`](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.current), or even in `PostgresStorage::new_async()` so that it doesn't need to be passed around.

File: core/lib/zksync_core/src/state_keeper/batch_executor/tests/read_storage_factory.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/batch_executor/tests/read_storage_factory.rs:27]: This looks incorrect; looks like you've fixed snapshot-related logic in the main code, but not here. Wouldn't it be possible to reuse the code from `AsyncRocksdbCache` here and in `impl ReadStorageFactory for RocksdbFactory` so that such mistakes are eliminated by design?

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:19]: Nit: The lifetime can be elided here; Rust assumes that the output borrows from `self` by default.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:100]: Nit: I'd consider returning `get_miniblock_range_of_l1_batch` returning `Ok(None)` an error here because the L1 batch is guaranteed to exist. So, it would make sense to convert it to an error using `.context(_)`.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:123]: Nit: You can use `anyhow::Context::context()` on `Option`s as well, so this can be shortened to `.context("Could not find snapshot, no state available")`.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:93]: Nit: May make sense to add `context` to DAL calls to understand which call has errored.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:167]: Bikshedding: `if let Some(_) = self.rocksdb_cell.get() { .. } else { .. }` is slightly shorter and sometimes considered more idiomatic.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:198]: I'd return `AsyncCatchupTask` from this method so that the caller can manage it. E.g., it'd make sense to add it to the list of managed tasks on the main node / EN (rn you'd need to call [`pending().await`](https://doc.rust-lang.org/std/future/fn.pending.html) so that when the task is logically complete, it doesn't bring the node down with it).

Comment (itegulov) [reply core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:198]: I used `stop_receiver.0.changed().await?` as that seemed like a slightly better alternative to `pending().await`

File: core/lib/state/src/utils.rs
Comment (itegulov) [reply core/lib/state/src/utils.rs]: Yeah, I managed to return RocksDB-related code to almost what it used to be before this PR.

Also, I got rid of all `From` implementations as they didn't fit the official `From` [guidelines](https://doc.rust-lang.org/std/convert/trait.From.html). Props to @popzxc for his Slack post about this the other day

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (itegulov) [reply core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs]: Oh, nice, had no idea it worked like that

**Review by RomanBrodetski — APPROVED** [review]
Looks great! Please wait for @slowli 's final approval before merging

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:17]: I wonder whether it would make sense to have a more defensive interface that takes the executed L1 batch number as an argument. Could be dealt with later, I suppose.

File: core/node/node_framework/src/implementations/layers/state_keeper/main_batch_executor.rs
Comment (slowli) [inline core/node/node_framework/src/implementations/layers/state_keeper/main_batch_executor.rs:56]: Nit: Missing `Debug` implementation and an blank line before the type declaration.

**Review by slowli — COMMENTED** [review]
Looks good; I only have minor nits other than spawning the syncing task.

File: core/lib/dal/src/blocks_dal.rs
Comment (itegulov) [reply core/lib/dal/src/blocks_dal.rs]: Good catch, no idea why I wrote "no" lol

File: core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs
Comment (itegulov) [reply core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs]: I imported it because I had to implement `Debug` for some custom type manually in the earlier stages of PR. But good catch that we don't need it anymore.

File: core/bin/external_node/src/main.rs
Comment (itegulov) [reply core/bin/external_node/src/main.rs]: Ahh, I didn't realize that's what you meant by "managed task". I assumed only node framework is capable of that. Should be better now!

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (itegulov) [reply core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:17]: Hmm, defensive in the sense that it would sanity check that the current state of PG matches the latest executed L1 batch or in a sense that it would force `AsyncRocksdbCache` to cut off state at that specific L1 batch somehow?

File: core/lib/zksync_core/src/block_reverter/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/block_reverter/mod.rs:224]: Nit: Not sure this change is warranted; the log above talks about the state keeper cache.

File: core/lib/zksync_core/src/state_keeper/batch_executor/tests/tester.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/batch_executor/tests/tester.rs:103]: Nit: It's unlikely that this type is changed in the future, so you can implement `Copy` for it and pass it by value.

File: core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs
Comment (slowli) [reply core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs:17]: The first option, i.e. a sanity check.

**Review by slowli — APPROVED** [review]
Very nice.

**Review by slowli — APPROVED** [review]

**Pull request merged**
